### ISSUE TITLE
Container to host networking not working with podman

### ISSUE THREAD

**Issue by atrauzzi** [dotnet/aspire#6846]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

Something doesn't seem right with how Aspire expects dapr sidecars to be visible for containers.

My dapr publisher is unable to reach the sidecar that Aspire configures for me:

```
io.netty.channel.ConnectTimeoutException: connection timed out: host.containers.internal/169.254.1.2:38047
```
_(this exception means that my extension was working up until the point it tries to reach out over the network)_

### Expected Behavior

I expect my container to be able to see the dapr sidecar.

### Steps To Reproduce

Regardless of whether I use `DOTNET_ASPIRE_CONTAINER_RUNTIME=podman`, or not, along with the following configuration:

```csharp
var daprDefaults = new DaprSidecarOptions
{
    RuntimePath = "./resources",
};

builder
    .AddKeycloak("auth", 8080, adminUser, adminPassword)
    .WithDataVolume()
    .WithBindMount("./resources/keycloak-dapr-publisher-0.0.1-all.jar", "/opt/keycloak/providers/keycloak-dapr-publisher-0.0.1-all.jar")
    .WithDaprSidecar(daprDefaults)
;
```

### Exceptions (if any)

No exceptions are reported by Aspire, although in my specific scenario above, the code depending on the dapr sidecar is unable to reach it, and ends up throwing an exception.

I know for a fact that this code works in a manually assembled scenario, so it's not a bug in that code.  You can also see that it specifically fails during resolution of the GRPC endpoint that Aspire injects on behalf of the dapr sidecar.

### .NET Version info

All 9s!

### Anything else?

In this case, I haven't done any network level customization, so Aspire is totally in the drivers seat.  But it's still difficult to tell when `host.containers.internal` is going to work and when it won't.

**Comment by davidfowl** [comment]
Yes this is container to host networking problems

**Comment by davidfowl** [comment]
If you can figure out what the issue is it would help us solve the issue 😊. Our plan for solving this is much more complex.

**Comment by davidfowl** [comment]
This doc is required reading for understanding networking in aspire https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/networking-overview.

I think the main question is, what host name should be used to contact the host from the container network (the port should be correct)

**Comment by davidfowl** [comment]
The port is correct. Read that document I linked, the proxy should be linked to the target (there's a proxy by default)

**Comment by davidfowl** [comment]
There is no proxy process, it's a single process (dcp) opening lots of ports on the host.

**Comment by davidfowl** [comment]
When you ran nc, what host did you use to connect to the host from the container ?

**Comment by davidfowl** [comment]
You can turning off the proxy and seeing if that works

**Comment by davidfowl** [comment]
That won't work as the sidecar resource is not the exact same resource as the CLI resource:

https://github.com/dotnet/aspire/blob/be904514949ba5d94e2e4b6d7b8890ab63b88316/src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs#L174-L176

**Comment by davidfowl** [comment]
@danegsta thoughts?

**Comment by danegsta** [comment]
The behavior of host.docker.internal/host.containers.internal depends on the specific OS and container runtime implementation.

With Linux, the host gateways are implemented via a custom network interface, which means traffic from Docker networking is technically a separate network than the host machine network. We made the decision early in Aspire to run our proxies bound to localhost in order to limit traffic to the local machine, but this means that they can't accept traffic routed via container host gateways. On Linux, any host ports accessible from Docker/Podman need to be either bound to all network interfaces (0.0.0.0) or specifically to the IP of the container gateway network interface.

Running on Windows, the problem is even more complicated, as Docker/Podman are running in a WSL instance, which adds an additional layer of network abstraction. The container host gateway networks route to WSL network interfaces, not to the default Windows network interface. Effectively the container networks can only access ports bound from the WSL side of the network. MacOS has similar behavior since the container runtime is running in a VM which has a similar level of network isolation as WSL on Windows.

Docker Desktop is currently the one scenario where things just work when routing traffic to the local network, as Docker Desktop implements the container host gateway in such a way that it directly routes traffic to localhost on the host machine (even when running in WSL on Windows) rather than via a separate network interface.

**Comment by danegsta** [comment]
TLDR is that we bind the proxy to localhost on the host machine due to security concerns with binding to 0.0.0.0, but that means they can't be reached via the host gateway from a container in most cases, due to the host gateway being implemented as a separate network interface.

**Comment by davidfowl** [comment]
I want [this comment](https://github.com/dotnet/aspire/issues/6846#issuecomment-2596671534) in known issues @IEvangelist.

@danegsta if we turn the proxy off this should "work" in this scenario right?

**Comment by danegsta** [comment]
> I want [this comment](https://github.com/dotnet/aspire/issues/6846#issuecomment-2596671534) in known issues [@IEvangelist](https://github.com/IEvangelist).
> 
> [@danegsta](https://github.com/danegsta) if we turn the proxy off this should "work" in this scenario right?

If the core service is binding to all interfaces, it's at least more likely to work.

**Comment by nielsNocore** [comment]
did anyone found a way to get podman working with containers to host network working.

**Comment by davidfowl** [comment]
I don't think we understand what is happening to provide workaround.
